packages/grafana-data/src/field/fieldState.ts

```diff
@@ -38,14 +38,45 @@ export function getFieldDisplayName(field: Field, frame?: DataFrame, allFrames?:
   }
 
   const displayName = calculateFieldDisplayName(field, frame, allFrames);
-  field.state = {
-    ...field.state,
-    displayName,
-  };
+  field.state = field.state || {};
+  field.state.displayName = displayName;
 
   return displayName;
 }
 
+/**
+ * Will keep the first occurance of all names, but clear any cached duplicate names
+ *
+ * This will trigger re-calculating the displayName when it is requested.
+ *
+ * NOTE: the field state may be modified
+ */
+export function cleanDuplicateFieldNames(frames: DataFrame[]): DataFrame[] {
+  const names = new Map<string, number>();
+  for (const frame of frames) {
+    names.clear();
+    for (const field of frame.fields) {
+      const name = field.state?.displayName ?? field.config.displayName;
+      if (name) {
+        const count = names.get(name);
+        if (count) {
+          // Modify display for duplicate configurations
+          if (field.config.displayName) {
+            field.state = field.state ?? {};
+            field.state!.displayName = `${field.config.displayName} ${count + 1}`;
+            names.set(name, count + 1);
+          } else {
+            field.state!.displayName = undefined;
+          }
+        } else {
+          names.set(name, 1);
+        }
+      }
+    }
+  }
+  return frames;
+}
+
```

Comment:
Looks good, think this is already handled by getFieldDisplayName 
https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/field/fieldState.ts#L136 

So another option could be to clear any cached display name?

Reply:
This also needs to run between every transform not just at the end